### PR TITLE
Send callbacks using server certificate

### LINKED ISSUE #10: Use server certificate when sending callback data

### PR DESCRIPTION
**Pull request by taubel**
Closes #10

### CODE REVIEW

**Comment by taubel** [comment]
Travis fails because currently the project uses an older version of ulfius which _doesn't support https_ (or has it off by default)

File: src/restserver.h [hunk 1]
```diff
@@ -52,6 +53,9 @@ typedef struct
 
     // rest-subsciptions
     rest_list_t *observeList;
+
+    char *certificate;
```
Comment (GiedriusM) [inline src/restserver.h]: Just save the reference to settings. Will be able to access information in future more easily.

File: src/rest-core.c [hunk 2]
```diff
@@ -91,6 +94,18 @@ int rest_step(rest_context_t *rest, struct timeval *tv)
         request.http_verb = strdup("PUT");
         request.http_url = strdup(url);
         request.timeout = 20;
+        request.check_server_certificate = 0;
+        request.client_cert_file = o_strdup(rest->certificate);
```
[the remark this answers is not in the document]
Comment (taubel) [reply src/rest-core.c]: o_strdup is recommended in ulfius examples. It also checks if 'rest->certificate' is NULL, and doesn't crash if so.

File: src/rest-core.c [hunk 3]
```diff
@@ -91,6 +92,18 @@ int rest_step(rest_context_t *rest, struct timeval *tv)
         request.http_verb = strdup("PUT");
         request.http_url = strdup(url);
         request.timeout = 20;
+        request.check_server_certificate = 0;
+        request.client_cert_file = o_strdup(rest->settings->http.security.certificate);
+        if(rest->settings->http.security.certificate != NULL && request.client_cert_file == NULL)
+        {
+            log_message(LOG_LEVEL_WARN, "[CALLBACK] Failed to set client certificate\n");
```
Comment (GiedriusM) [inline src/rest-core.c]: If there's a failure in request configuration it should not be sent.

Comment (taubel) [reply src/rest-core.c]: Added abort sequence, please check if it is correct

**Comment by GiedriusM** [comment]
Can you test https://github.com/punica/ulfius/tree/pr-separate-gnutls? It should have separate support for client certificate (WITH_GNUTLS=on) without websocket dependency (WITH_WEBSOCKETS=off).

**Comment by taubel** [comment]
Open for suggestions for additional tests

File: script/bootstrap [hunk 4]
```diff
@@ -133,7 +133,7 @@ install_libjwt () {
 
 install_ulfius_orcania_yder () {
     ULFIUS_DIR="${TMP_BUILD_DIR}/ulfius"
-    git clone --quiet -b v2.5.2 --single-branch https://github.com/babelouest/ulfius.git "${ULFIUS_DIR}"
+    git clone --quiet -b pr-separate-gnutls --single-branch https://github.com/punica/ulfius.git "${ULFIUS_DIR}"
```
Comment (GiedriusM) [inline script/bootstrap]: Changes are now available in ulfius v2.5.3, except currently there's no tag, so just use commit number `78e35168d3483050c01e868a681d40d7fa9de801`

**Pull request merged**